File: python/searching-and-sorting/searching/binary-search.py
def binarySearch(arr, l, r, x):
    if r >= l:
        mid = l + (r - l) // 2
        if arr[mid] == x:
            return mid
        elif arr[mid] > x:
            return binarySearch(arr, l, mid - 1, x)
        else:
            return binarySearch(arr, mid + 1, r, x)
    else:
        return -1

arr = [1, 2, 3, 4, 5, 6]
x = 6
n = len(arr)

result = binarySearch(arr, 0, n-1, x)
if result == -1:
    print("Element is not present in array")
else:
    print("Element is present at index", result)

File: cpp/searching-and-sorting/searching/binary-search.cpp
#include<conio.h>
#include<iostream>
using namespace std;

int search(int arr[], int n, int x) {
	int mid;
	int left = 0, right = n - 1;
	while(left != right) {
		mid = (left + right) / 2;
		if(x > arr[mid])	left = mid + 1;
		else right = mid - 1;
	}
	if(arr[left] == x) return left;
	return -1;
}

int binarySearchRecursive(int arr[], int l, int r, int x) {
	if (r >= l) {
		int mid = l + (r - l) / 2;
		if (arr[mid] == x)
			return mid;
		if (arr[mid] > x)
			return binarySearchRecursive(arr, l, mid - 1, x);
		return binarySearchRecursive(arr, l + 1, r, x);
	}
	return -1;
}

int main() {
	int arr[] = {1, 2, 3, 4, 5, 6};
	int n = sizeof(arr) / sizeof(arr[0]);
	int x = 6;
	int result = binarySearchRecursive(arr, 0, n - 1, x);
	(result == -1)
		? cout<<"Element is not present in array"
		: cout<<"Element is present in array at index " << result;
	
	int result2 = search(arr, n , x);
	(result2 == -1)
		? cout<<"Element is not present in array"
		: cout<<"Element is present in array at index " << result2;
	return 0;
}
